# Základní dotazy

<https://kodim.cz/kurzy/python-data-1/python-pro-data-1/zakladni-dotazy/zakladni-dotazy>



In [2]:
# Stahneme si data, s kterymi budeme pracovat
import requests

url = "https://kodim.cz/cms/assets/kurzy/python-data-1/python-pro-data-1/zakladni-dotazy/staty.json"
r = requests.get(url) # <Response [200]>
open("staty.json", "wb").write(r.content)

42534

## Indexy

### Přidání indexů

In [3]:
import pandas

staty = pandas.read_json("../soubory_k_praci/staty.json")
staty.head() #staty.info


name alpha2Code alpha3Code    capital   region        subregion  \
0     Afghanistan         AF        AFG      Kabul     Asia    Southern Asia   
1   Åland Islands         AX        ALA  Mariehamn   Europe  Northern Europe   
2         Albania         AL        ALB     Tirana   Europe  Southern Europe   
3         Algeria         DZ        DZA    Algiers   Africa  Northern Africa   
4  American Samoa         AS        ASM  Pago Pago  Oceania        Polynesia   

   population       area  gini  
0    27657145   652230.0  27.8  
1       28875     1580.0   NaN  
2     2886026    28748.0  34.5  
3    40400000  2381741.0  35.3  
4       57100      199.0   NaN

Každý stát na světě má svůj název a ten název je unikátní a identifikuje ho. 

Můžeme tedy tento název použít jako index.

To, jaký sloupec má být použit jako index, řeší funkce `set_index()`.

In [4]:
# Nastaveni indexu
staty = staty.set_index("name")

Zkontrolovat index můžeme pomocí příkazu `staty.index`.

In [5]:
staty.index

Index(['Afghanistan', 'Åland Islands', 'Albania', 'Algeria', 'American Samoa',
       'Andorra', 'Angola', 'Anguilla', 'Antarctica', 'Antigua and Barbuda',
       ...
       'Uruguay', 'Uzbekistan', 'Vanuatu',
       'Venezuela (Bolivarian Republic of)', 'Viet Nam', 'Wallis and Futuna',
       'Western Sahara', 'Yemen', 'Zambia', 'Zimbabwe'],
      dtype='object', name='name', length=250)

### Výběr konkrétního řádku a hodnoty

K nalezení řádku pomocí indexu použijeme `loc`, která funguje obdobně jako funkce `iloc`. 

* `iloc` pracuje s čísly řádků
* `loc` používá nami zvolené indexy

In [6]:
# vyber radku
staty.loc["Czech Republic"]

alpha2Code                CZ
alpha3Code               CZE
capital               Prague
region                Europe
subregion     Eastern Europe
population          10558524
area                 78865.0
gini                    26.0
Name: Czech Republic, dtype: object

In [7]:
# vyber radku a sloupce
staty.loc["Czech Republic","area"]

78865.0

### Výběr několika řádků

In [8]:
# Náš soubor je seřazený dle abecedy.
staty.loc["Czech Republic":"Dominican Republic"]

# staty.loc[:"Andorra"]
# staty.loc["Uzbekistan":] 
# staty.loc[["Czech Republic","Slovakia"]]
# staty.loc[["Slovakia","Poland","Germany","Austria"], ["area","population"]]

alpha2Code alpha3Code        capital    region  \
name                                                                
Czech Republic             CZ        CZE         Prague    Europe   
Denmark                    DK        DNK     Copenhagen    Europe   
Djibouti                   DJ        DJI       Djibouti    Africa   
Dominica                   DM        DMA         Roseau  Americas   
Dominican Republic         DO        DOM  Santo Domingo  Americas   

                          subregion  population     area  gini  
name                                                            
Czech Republic       Eastern Europe    10558524  78865.0  26.0  
Denmark             Northern Europe     5717014  43094.0  24.0  
Djibouti             Eastern Africa      900000  23200.0  40.0  
Dominica                  Caribbean       71293    751.0   NaN  
Dominican Republic        Caribbean    10075045  48671.0  47.2

## Dotazy

Série umí například sama spočítat svůj součet a vrátit výsledek jako číslo. 

Dokumentace: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html

In [9]:
populace = staty["population"]
populace.sum()

7349137231

In [10]:
print(f"Prumer je {populace.mean()}.")
print(f"Minumum je {populace.min()}.")
print(f"Maximum je {populace.max()}.")
print(f"Median je {populace.median()}.")

Prumer je 29396548.924.
Minumum je 0.
Maximum je 1377422166.
Median je 4648733.5.


In [11]:
populace.describe()

# staty.describe() # lze pouzit na cely DataFrame

count    2.500000e+02
mean     2.939655e+07
std      1.244578e+08
min      0.000000e+00
25%      2.029030e+05
50%      4.648734e+06
75%      1.803955e+07
max      1.377422e+09
Name: population, dtype: float64

## Podmínky

Ověřením podmínky získáme sloupec, kde máme 250 hodnot typu `bool`.

In [12]:
# státy, které mají méně než 1000 obyvatel

staty["population"] < 1000


name
Afghanistan          False
Åland Islands        False
Albania              False
Algeria              False
American Samoa       False
                     ...  
Wallis and Futuna    False
Western Sahara       False
Yemen                False
Zambia               False
Zimbabwe             False
Name: population, Length: 250, dtype: bool

`pandas` teď jednoduše udělají to, že vypíšou ty řádky řádky, kde má náš polotovar hodnotu `True` a ty, které mají hodnotu `False`, před námi skryjí.

In [13]:
pidistaty = staty[staty["population"] < 1000]

pidistaty[["population", "area"]] 

population     area
name                                                             
Bouvet Island                                          0    49.00
United States Minor Outlying Islands                 300      NaN
Cocos (Keeling) Islands                              550    14.00
French Southern Territories                          140  7747.00
Heard Island and McDonald Islands                      0   412.00
Holy See                                             451     0.44
Pitcairn                                              56    47.00
South Georgia and the South Sandwich Islands          30      NaN

V některých řádcích vidíme hodnotu NaN. To značí, že pro daný řádek hodnotu nemáme, pro některé státy tedy nemáme zadanou rozlohu. 

In [14]:
# Znas SQL? vyzkousej funkci query()
staty.query("population < 1000")

alpha2Code alpha3Code  \
name                                                                 
Bouvet Island                                        BV        BVT   
United States Minor Outlying Islands                 UM        UMI   
Cocos (Keeling) Islands                              CC        CCK   
French Southern Territories                          TF        ATF   
Heard Island and McDonald Islands                    HM        HMD   
Holy See                                             VA        VAT   
Pitcairn                                             PN        PCN   
South Georgia and the South Sandwich Islands         GS        SGS   

                                                        capital    region  \
name                                                                        
Bouvet Island                                                               
United States Minor Outlying Islands                             Americas   
Cocos (Keeling) Islands                             West Island   Oceania   
French Southern Territories                   Port-aux-Français    Africa   
Heard Island and McDonald Islands                                           
Holy See                                                   Rome    Europe   
Pitcairn                                              Adamstown   Oceania   
South Georgia and the South Sandwich Islands  King Edward Point  Americas   

                                                              subregion  \
name                                                                      
Bouvet Island                                                             
United States Minor Outlying Islands                   Northern America   
Cocos (Keeling) Islands                       Australia and New Zealand   
French Southern Territories                             Southern Africa   
Heard Island and McDonald Islands                                         
Holy See                                                Southern Europe   
Pitcairn                                                      Polynesia   
South Georgia and the South Sandwich Islands              South America   

                                              population     area  gini  
name                                                                     
Bouvet Island                                          0    49.00   NaN  
United States Minor Outlying Islands                 300      NaN   NaN  
Cocos (Keeling) Islands                              550    14.00   NaN  
French Southern Territories                          140  7747.00   NaN  
Heard Island and McDonald Islands                      0   412.00   NaN  
Holy See                                             451     0.44   NaN  
Pitcairn                                              56    47.00   NaN  
South Georgia and the South Sandwich Islands          30      NaN   NaN

### Spojení více podmínek

Pokud chceme, aby musely být splněny obě podmínky, vložíme mezi ně symbol `&`. 

Pokud chceme, aby stačilo splnění jedné podmínky, použijeme symbol `|`. 



In [15]:
lidnate_evropske_staty = staty[(staty["population"] > 20_000_000) & (staty["region"] == "Europe")]
lidnate_evropske_staty["population"]

name
France                                                   66710000
Germany                                                  81770900
Italy                                                    60665551
Poland                                                   38437239
Russian Federation                                      146599183
Spain                                                    46438422
Ukraine                                                  42692393
United Kingdom of Great Britain and Northern Ireland     65110000
Name: population, dtype: int64

In [16]:
staty[(staty["population"] > 10_000_000_000) | (staty["area"] > 3_000_000)]

alpha2Code alpha3Code           capital    region  \
name                                                                         
Antarctica                       AQ        ATA                       Polar   
Australia                        AU        AUS          Canberra   Oceania   
Brazil                           BR        BRA          Brasília  Americas   
Canada                           CA        CAN            Ottawa  Americas   
China                            CN        CHN           Beijing      Asia   
India                            IN        IND         New Delhi      Asia   
Russian Federation               RU        RUS            Moscow    Europe   
United States of America         US        USA  Washington, D.C.  Americas   

                                          subregion  population        area  \
name                                                                          
Antarctica                                                 1000  14000000.0   
Australia                 Australia and New Zealand    24117360   7692024.0   
Brazil                                South America   206135893   8515767.0   
Canada                             Northern America    36155487   9984670.0   
China                                  Eastern Asia  1377422166   9640011.0   
India                                 Southern Asia  1295210000   3287590.0   
Russian Federation                   Eastern Europe   146599183  17124442.0   
United States of America           Northern America   323947000   9629091.0   

                          gini  
name                            
Antarctica                 NaN  
Australia                 30.5  
Brazil                    54.7  
Canada                    32.6  
China                     47.0  
India                     33.4  
Russian Federation        40.1  
United States of America  48.0

### Použití seznamu v podmínce

Funkce `isin()` aplikovaná na jeden konkrétní sloupec vrátí nám `True` pro všechny řádky, pro které je hodnota přítomná v seznamu. 

In [17]:
staty[staty["subregion"].isin(["Western Europe", "Eastern Europe"])]

alpha2Code alpha3Code     capital  region  \
name                                                              
Austria                       AT        AUT      Vienna  Europe   
Belarus                       BY        BLR       Minsk  Europe   
Belgium                       BE        BEL    Brussels  Europe   
Bulgaria                      BG        BGR       Sofia  Europe   
Czech Republic                CZ        CZE      Prague  Europe   
France                        FR        FRA       Paris  Europe   
Germany                       DE        DEU      Berlin  Europe   
Hungary                       HU        HUN    Budapest  Europe   
Liechtenstein                 LI        LIE       Vaduz  Europe   
Luxembourg                    LU        LUX  Luxembourg  Europe   
Moldova (Republic of)         MD        MDA    Chișinău  Europe   
Monaco                        MC        MCO      Monaco  Europe   
Netherlands                   NL        NLD   Amsterdam  Europe   
Poland                        PL        POL      Warsaw  Europe   
Republic of Kosovo            XK        KOS    Pristina  Europe   
Romania                       RO        ROU   Bucharest  Europe   
Russian Federation            RU        RUS      Moscow  Europe   
Slovakia                      SK        SVK  Bratislava  Europe   
Switzerland                   CH        CHE        Bern  Europe   
Ukraine                       UA        UKR        Kiev  Europe   

                            subregion  population         area  gini  
name                                                                  
Austria                Western Europe     8725931     83871.00  26.0  
Belarus                Eastern Europe     9498700    207600.00  26.5  
Belgium                Western Europe    11319511     30528.00  33.0  
Bulgaria               Eastern Europe     7153784    110879.00  28.2  
Czech Republic         Eastern Europe    10558524     78865.00  26.0  
France                 Western Europe    66710000    640679.00  32.7  
Germany                Western Europe    81770900    357114.00  28.3  
Hungary                Eastern Europe     9823000     93028.00  31.2  
Liechtenstein          Western Europe       37623       160.00   NaN  
Luxembourg             Western Europe      576200      2586.00  30.8  
Moldova (Republic of)  Eastern Europe     3553100     33846.00  33.0  
Monaco                 Western Europe       38400         2.02   NaN  
Netherlands            Western Europe    17019800     41850.00  30.9  
Poland                 Eastern Europe    38437239    312679.00  34.1  
Republic of Kosovo     Eastern Europe     1733842     10908.00   NaN  
Romania                Eastern Europe    19861408    238391.00  30.0  
Russian Federation     Eastern Europe   146599183  17124442.00  40.1  
Slovakia               Eastern Europe     5426252     49037.00  26.0  
Switzerland            Western Europe     8341600     41284.00  33.7  
Ukraine                Eastern Europe    42692393    603700.00  26.4

## Převody dat na DataFrame a zpět

### Převod DataFrame na seznam

K takovému převodu na seznam nám poslouží kombinace funkcí `to_numpy` a `tolist`. 

Převod totiž neprovádíme přímo, ale jako mezikrok jej převedeme na pole modulu `numpy`.

In [18]:
staty_list = staty.to_numpy().tolist() 
staty_list[0] 

['AF', 'AFG', 'Kabul', 'Asia', 'Southern Asia', 27657145, 652230.0, 27.8]

Index se v Pandas nebere jako součást dat. 

Pokud chceme index vrátit do původního stavu a mít ho jako automaticky generovaná čísla řádků, můžeme použít metodu `reset_index`.

In [19]:
staty_list = staty.reset_index().to_numpy().tolist() 

staty_list[0]

['Afghanistan',
 'AF',
 'AFG',
 'Kabul',
 'Asia',
 'Southern Asia',
 27657145,
 652230.0,
 27.8]

### Vytvoření DataFrame ze seznamu

In [20]:
znamky = [
    ['Petr', 2],
    ['Roman', 1],
    ['Jitka', 3],
    ['Zuzana', 5],
    ['Ondřej', 2],
    ['Julie', 2],
    ['Karel', 4],
    ['Anna', 1],
    ['Eva', 1]
]

soucet = 0
for radek in znamky:
    soucet = soucet + radek[1]
prumer = soucet / len(znamky)

In [28]:
znamky = pandas.DataFrame(znamky, columns=['student', 'znamka'])

prumer = znamky["znamka"].mean()

2.3333333333333335

### Vytvoření DataFrame ze seznamu slovníků

Nemusíme přidávat názvy sloupců, protože ty už funkce DataFrame získá z klíčů slovníků.

In [22]:
nakupy = [
    {"person": "Petr", "item": "Prací prášek", "value": 399},
    {"person": "Ondra", "item": "Savo", "value": 80},
    {"person": "Petr", "item": "Toaletní papír", "value": 65},
    {"person": "Libor", "item": "Pivo", "value": 124},
    {"person": "Petr", "item": "Pytel na odpadky", "value": 75},
    {"person": "Míša", "item": "Utěrky na nádobí", "value": 130},
    {"person": "Ondra", "item": "Toaletní papír", "value": 120},
    {"person": "Míša", "item": "Pečící papír", "value": 30},
    {"person": "Zuzka", "item": "Savo", "value": 80},
    {"person": "Pavla", "item": "Máslo", "value": 50},
    {"person": "Ondra", "item": "Káva", "value": 300}
]

In [25]:
nakupy = pandas.DataFrame(nakupy)
nakupy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11 entries, 0 to 10
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   person  11 non-null     object
 1   item    11 non-null     object
 2   value   11 non-null     int64 
dtypes: int64(1), object(2)
memory usage: 392.0+ bytes
